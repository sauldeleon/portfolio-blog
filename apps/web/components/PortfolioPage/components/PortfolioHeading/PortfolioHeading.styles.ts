import Image from 'next/image'
import { styled } from 'styled-components'

import { Body, Heading, Label } from '@sdlgr/typography'

export const StyledSection = styled.section`
  border: 3px solid ${({ theme }) => theme.colors.white};
  border-bottom: none;
  padding: 1rem;
  width: 100%;
  display: flex;
  flex-direction: column;

  ${({ theme }) => theme.media.up.lg} {
    padding: 3rem;
  }
`

export const StyledRow = styled.div`
  display: flex;
  width: 100%;
  justify-content: space-between;
  align-items: flex-start;
`

export const StyledDegree = styled(Label)`
  font-size: 16px;
  margin-bottom: 20px;
  width: 100%;

  ${({ theme }) => theme.media.up.md} {
    font-size: 20px;
  }
`

export const StyledFirstRow = styled(StyledRow)`
  flex-direction: column-reverse;
  font-size: 16px;

  ${({ theme }) => theme.media.up.md} {
    flex-direction: row;
    font-size: 20px;
  }
`

export const StyledInfo = styled.div`
  text-transform: uppercase;
  margin-bottom: 30px;
  width: 100%;
  text-align: right;

  ${({ theme }) => theme.media.up.md} {
    margin-bottom: 50px;
  }
`

export const StyledBody = styled(Body)`
  font-size: 16px;

  ${({ theme }) => theme.media.up.lg} {
    font-size: 20px;
  }
`

export const StyledPortrait = styled(Image)`
  overflow: hidden;
  border-radius: 50%;
  width: 60px;
  min-width: 60px;
  height: 60px;

  ${({ theme }) => theme.media.up.md} {
    width: 180px;
    min-width: 180px;
    height: 180px;
  }

  ${({ theme }) => theme.media.up.lg} {
    width: 310px;
    min-width: 310px;
    height: 310px;
  }
`

export const StyledHeading = styled(Heading)`
  margin-top: 25px;
  margin-left: 20px;
  font-size: 1.5rem;
  font-weight: 500;

  ${({ theme }) => theme.media.up.md} {
    margin-top: 40px;
    margin-top: 100px;
    margin-left: 40px;
    font-size: 3rem;
    font-weight: 500;
  }

  ${({ theme }) => theme.media.up.lg} {
    margin-top: 160px;
    margin-left: 80px;
    font-size: 4rem;
    font-weight: 500;
  }
`
