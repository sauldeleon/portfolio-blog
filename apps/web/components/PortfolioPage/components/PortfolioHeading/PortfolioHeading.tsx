import { useClientTranslation } from '@web/i18n/client'

import {
  StyledBody,
  StyledDegree,
  StyledFirstRow,
  StyledHeading,
  StyledInfo,
  StyledPortrait,
  StyledRow,
  StyledSection,
} from './PortfolioHeading.styles'

export function PortfolioHeading() {
  const { t } = useClientTranslation('portfolioPage')
  return (
    <StyledSection>
      <StyledFirstRow>
        <StyledDegree>{t('items.softwareEngineer')}</StyledDegree>
        <StyledInfo>
          <StyledBody>{t('basedIn')}</StyledBody>
          <StyledBody>sauldeleonguerrero@gmail.com</StyledBody>
        </StyledInfo>
      </StyledFirstRow>
      <StyledRow>
        <StyledPortrait
          src="/assets/portrait-4.png"
          width={310}
          height={310}
          alt={t('profilePicture')}
        />
        <StyledHeading $level={1}>Saúl de León Guerrero</StyledHeading>
      </StyledRow>
    </StyledSection>
  )
}
