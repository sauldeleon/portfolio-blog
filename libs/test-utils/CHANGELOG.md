# Changelog Test utils

## [0.2.3](https://github.com/sauldeleon/portfolio-blog/compare/test-utils-0.2.2...test-utils-0.2.3) (2024-01-08)

## [0.2.2](https://github.com/sauldeleon/portfolio-blog/compare/test-utils-0.2.1...test-utils-0.2.2) (2023-09-26)

## [0.2.1](https://github.com/sauldeleon/portfolio-blog/compare/test-utils-0.2.0...test-utils-0.2.1) (2023-09-14)

# [0.2.0](https://github.com/sauldeleon/portfolio-blog/compare/test-utils-0.1.0...test-utils-0.2.0) (2023-09-05)

### Bug Fixes

- reduced duplicated lines ([f51168a](https://github.com/sauldeleon/portfolio-blog/commit/f51168aefe025a5a65ffdb2dc22f7a16b863406e))
- rename commit message format ([e1e2ab4](https://github.com/sauldeleon/portfolio-blog/commit/e1e2ab404bbd2c32f3508d1ed8197b3fbff93cb9))

### Features

- deleted changelogs ([e9be338](https://github.com/sauldeleon/portfolio-blog/commit/e9be33836ee47b6505ad94d21f4be21855a7fa0d))

# 0.1.0 (2023-09-02)

### Bug Fixes

- rename commit message format ([e1e2ab4](https://github.com/sauldeleon/portfolio-blog/commit/e1e2ab404bbd2c32f3508d1ed8197b3fbff93cb9))

### Features

- add contact initial page ([fea862b](https://github.com/sauldeleon/portfolio-blog/commit/fea862b3507268013149050ae3f05fa517e7b3d9))
- add homepage ([60523aa](https://github.com/sauldeleon/portfolio-blog/commit/60523aa31ec088d318745ef0c757b4cbe656f0d4))
- deleted changelogs ([e9be338](https://github.com/sauldeleon/portfolio-blog/commit/e9be33836ee47b6505ad94d21f4be21855a7fa0d))
- improved a11y and styles ([873f656](https://github.com/sauldeleon/portfolio-blog/commit/873f656587f50df8e29ad526fdc68fea0b2c9784))
- toggle language button ([afd97e9](https://github.com/sauldeleon/portfolio-blog/commit/afd97e95beac0f70ed643c753e71db2ca014054b))
