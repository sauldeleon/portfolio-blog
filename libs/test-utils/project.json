{
  "name": "test-utils",
  "$schema": "../../node_modules/nx/schemas/project-schema.json",
  "sourceRoot": "libs/test-utils/src",
  "projectType": "library",
  "tags": [],
  "targets": {
    "version": {
      "executor": "@jscutlery/semver:version",
      "options": {
        "changelogHeader": "# Changelog Test utils \n",
        "baseBranch": "main",
        "commitMessageFormat": "chore(${projectName}): release: v${version}",
        "push": false
      }
    },
    "lint": {
      "executor": "@nx/linter:eslint",
      "outputs": ["{options.outputFile}"],
      "options": {
        "lintFilePatterns": ["libs/test-utils/**/*.{ts,tsx,js,jsx}"]
      }
    },
    "test": {
      "executor": "@nx/jest:jest",
      "outputs": ["{workspaceRoot}/coverage/{projectRoot}"],
      "options": {
        "jestConfig": "libs/test-utils/jest.config.ts",
        "passWithNoTests": true
      },
      "configurations": {
        "ci": {
          "ci": true,
          "codeCoverage": true
        }
      }
    }
  }
}
