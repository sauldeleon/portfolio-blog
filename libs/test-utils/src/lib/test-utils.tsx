import { QueryClient, QueryClientProvider } from '@tanstack/react-query'
import { Matcher } from '@testing-library/dom/types/matches'
import {
  RenderOptions,
  RenderResult,
  queryAllByAttribute as _queryAllByAttribute,
  queryByAttribute as _queryByAttribute,
  act,
  render,
  screen,
} from '@testing-library/react'
import { RouterContext } from 'next/dist/shared/lib/router-context'
import Router, { NextRouter } from 'next/router'
import { ReactNode, useEffect, useState } from 'react'
import { ThemeProvider } from 'styled-components'

import { LanguageContextProvider } from '@sdlgr/i18n-config'
import { mainTheme } from '@sdlgr/main-theme'

export type MatcherCustomFnParams = {
  content: string
  element: Element | null
}

/* istanbul ignore next */
export const renderWithTheme = (children: ReactNode): RenderResult => {
  return render(<ThemeProvider theme={mainTheme}>{children}</ThemeProvider>)
}

/* istanbul ignore next */
export function RenderProviders({ children }: { children: ReactNode }) {
  const queryClient = new QueryClient({
    defaultOptions: {
      queries: {
        retry: false,
      },
    },
  })

  return (
    <QueryClientProvider client={queryClient}>
      <LanguageContextProvider value={{ language: 'en' }}>
        <ThemeProvider theme={mainTheme}>{children}</ThemeProvider>
      </LanguageContextProvider>
    </QueryClientProvider>
  )
}

/* istanbul ignore next */
export const renderApp = (
  children: ReactNode,
  options?: Omit<RenderOptions, 'queries'>
) => {
  return render(<RenderProviders>{children}</RenderProviders>, options)
}

/**
 * Generic function to get element by attribute and value
 * @example queryByAttribute('id')(baseElement, 'element-id')
 * @param {string} attr
 * @return {any}
 */
export function queryByAttribute(attr: string) {
  return _queryByAttribute.bind(null, attr)
}

/**
 * Generic function to get all elements by attribute and value
 * @example queryByAttribute('class')(baseElement, 'element-class')
 * @param {string} attr
 * @returns {any}
 */
export function queryAllByAttribute(attr: string) {
  return _queryAllByAttribute.bind(null, attr)
}

/**
 * Generic function to get element by attribute and value
 * @example getByAttribute('id')(baseElement, 'element-id')
 * @param {string} attr
 * @return {any}
 */
export function getByAttribute(attr: string) {
  return _queryByAttribute.bind(null, attr)
}

/**
 * Generic function to get all elements by attribute and value
 * @example getByAttribute('class')(baseElement, 'element-class')
 * @param {string} attr
 * @returns {any}
 */
export function getAllByAttribute(attr: string) {
  return _queryAllByAttribute.bind(null, attr)
}

/**
 * Shortcut version for queryByAttribute, but specific for id
 * @param {HTMLElement} container
 * @param {string} id
 * @return {any}
 */
export function getById(container: HTMLElement, id: string) {
  return queryByAttribute('id')(container, id) as Element
}

/**
 * Shortcut version for queryByAttribute, but specific for class name
 * @param {HTMLElement} container
 * @param {string} className
 * @return {any}
 */
export function getByClassName(container: HTMLElement, className: string) {
  return queryByAttribute('class')(container, className) as Element
}

function matcherCustomFn(
  tag: string,
  matcher: Matcher,
  { element, content }: MatcherCustomFnParams
) {
  const isTag = element?.tagName.toUpperCase() === tag.toUpperCase()
  let matching = false
  if (isTag) {
    matching = content === matcher
    /* istanbul ignore else*/
    if (typeof matcher === 'function') {
      matching = matcher(content, element)
    } else if (matcher instanceof RegExp) {
      matching = matcher.test(content)
    }
  }
  return matching && isTag
}

/**
 * Get 1 element by tag and text
 * @param {string} tag
 * @param {Matcher} matcher
 * @return {any}
 */
export function getByTagAndText(tag: string, matcher: Matcher) {
  return screen.getByText((content, element) =>
    matcherCustomFn(tag, matcher, { content, element })
  )
}

/**
 * Get multiple elements by tag and text
 * @param {string} tag
 * @param {Matcher} matcher
 * @return {any}
 */
export function getAllByTagAndText(tag: string, matcher: Matcher) {
  return screen.getAllByText((content, element) =>
    matcherCustomFn(tag, matcher, { content, element })
  )
}

/**
 * Mock Math.random() to return a specific value
 * @param {number} value
 */
export function mockRandom(value = 0.1234) {
  const mockMath = Object.create(global.Math)
  mockMath.random = () => value
  global.Math = mockMath
}

/**
 * Hook to use in test to ensure "last" render is called
 */
export function useWithRefreshInTest() {
  const [withRefreshInTest, setWithRefreshInTest] = useState(false)

  useEffect(() => {
    // Change state to "rerender" the component
    setWithRefreshInTest(true)
  }, [])

  return withRefreshInTest
}

export { render } from '@testing-library/react'

/* istanbul ignore next */
export const awaitAllAsync = async () =>
  act(() => new Promise((resolve) => setTimeout(resolve, 0)))

/* istanbul ignore next */
export function renderWithNextRouter(
  children: ReactNode,
  options?: Partial<NextRouter>
) {
  return renderWithTheme(
    <RouterContext.Provider value={{ ...Router, ...options }}>
      {children}
    </RouterContext.Provider>
  )
}
