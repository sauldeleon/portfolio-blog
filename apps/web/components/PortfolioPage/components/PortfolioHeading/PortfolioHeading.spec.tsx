import { screen, waitFor } from '@testing-library/react'

import { renderApp } from '@sdlgr/test-utils'

import { PortfolioHeading } from './PortfolioHeading'

describe('PortfolioHeading', () => {
  it('should render correctly', async () => {
    const { baseElement } = renderApp(<PortfolioHeading />)
    await waitFor(() =>
      expect(
        screen.getByText('sauldeleonguerrero@gmail.com'),
      ).toBeInTheDocument(),
    )
    expect(baseElement).toMatchSnapshot()
  })
})
