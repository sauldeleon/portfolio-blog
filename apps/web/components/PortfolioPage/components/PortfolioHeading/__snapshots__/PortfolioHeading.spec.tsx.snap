// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PortfolioHeading should render correctly 1`] = `
.c9 {
  font-family: fontFamily;
  font-weight: 400;
  color: #FBFBFB;
  font-size: 3rem;
  line-height: normal;
  margin: 0;
}

.c6 {
  font-family: fontFamily;
  font-weight: 400;
  color: #FBFBFB;
  font-size: 0.9rem;
  line-height: normal;
  white-space: pre-line;
}

.c6 ul,
.c6 ol {
  margin-left: 40px;
  margin-bottom: 10px;
}

.c6 ul {
  list-style: outside disc;
}

.c6 ol {
  list-style: decimal;
}

.c6 li {
  padding-left: 10px;
}

.c3 {
  font-family: fontFamily;
  font-weight: 500;
  color: #FBFBFB;
  font-size: 1rem;
  line-height: normal;
  text-transform: uppercase;
}

.c0 {
  border: 3px solid #FBFBFB;
  border-bottom: none;
  padding: 1rem;
  width: 100%;
  display: flex;
  flex-direction: column;
}

.c1 {
  display: flex;
  width: 100%;
  justify-content: space-between;
  align-items: flex-start;
}

.c4 {
  font-size: 16px;
  margin-bottom: 20px;
  width: 100%;
}

.c2 {
  flex-direction: column-reverse;
  font-size: 16px;
}

.c5 {
  text-transform: uppercase;
  margin-bottom: 30px;
  width: 100%;
  text-align: right;
}

.c7 {
  font-size: 16px;
}

.c8 {
  overflow: hidden;
  border-radius: 50%;
  width: 60px;
  min-width: 60px;
  height: 60px;
}

.c10 {
  margin-top: 25px;
  margin-left: 20px;
  font-size: 1.5rem;
  font-weight: 500;
}

@media (width >= 1024px) {
  .c0 {
    padding: 3rem;
  }
}

@media (width >= 768px) {
  .c4 {
    font-size: 20px;
  }
}

@media (width >= 768px) {
  .c2 {
    flex-direction: row;
    font-size: 20px;
  }
}

@media (width >= 768px) {
  .c5 {
    margin-bottom: 50px;
  }
}

@media (width >= 1024px) {
  .c7 {
    font-size: 20px;
  }
}

@media (width >= 768px) {
  .c8 {
    width: 180px;
    min-width: 180px;
    height: 180px;
  }
}

@media (width >= 1024px) {
  .c8 {
    width: 310px;
    min-width: 310px;
    height: 310px;
  }
}

@media (width >= 768px) {
  .c10 {
    margin-top: 40px;
    margin-top: 100px;
    margin-left: 40px;
    font-size: 3rem;
    font-weight: 500;
  }
}

@media (width >= 1024px) {
  .c10 {
    margin-top: 160px;
    margin-left: 80px;
    font-size: 4rem;
    font-weight: 500;
  }
}

<body>
  <div>
    <section
      class="c0"
    >
      <div
        class="c1 c2"
      >
        <div
          class="c3 c4"
        >
          Software engineer
        </div>
        <div
          class="c5"
        >
          <p
            class="c6 c7"
          >
            Based in Madrid, Spain
          </p>
          <p
            class="c6 c7"
          >
            sauldeleonguerrero@gmail.com
          </p>
        </div>
      </div>
      <div
        class="c1"
      >
        <img
          alt="My profile picture"
          class="c8"
          data-nimg="1"
          decoding="async"
          height="310"
          loading="lazy"
          src="/_next/image?url=%2Fassets%2Fportrait-4.png&w=640&q=75"
          srcset="/_next/image?url=%2Fassets%2Fportrait-4.png&w=384&q=75 1x, /_next/image?url=%2Fassets%2Fportrait-4.png&w=640&q=75 2x"
          style="color: transparent;"
          width="310"
        />
        <h1
          class="c9 c10"
        >
          Saúl de León Guerrero
        </h1>
      </div>
    </section>
  </div>
</body>
`;
